File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day1/Acwing1497/Main.java
package Week2.day1.Acwing1497;

import java.util.*;

/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/25 16:14
 */
public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int a[] = new int[n];
        for (int i = 0; i < n; i++) {   //后序
            a[i] = sc.nextInt();
        }
        //中序
        int b[] = new int[n];


        for (int i = 0; i < n; i++)
            b[i] = sc.nextInt();
        Solution solution=new Solution();

        TreeNode root=solution.buildTree(a,b);
        Queue<TreeNode>v=new LinkedList<>();
        v.add(root);
        while (v.size()!=0){
            TreeNode treeNode=v.poll();
            System.out.printf("%d ",treeNode.val);
            if(treeNode.left!=null)
                v.add(treeNode.left);
            if(treeNode.right!=null)
                v.add(treeNode.right);
        }


    }

}
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        //为空直接跳出
        if (postorder == null || postorder.length == 0) {
            return null;
        }
        //总的根节点一定是后序遍历的最后一个节点
        TreeNode root = new TreeNode(postorder[postorder.length - 1]);
        Deque<TreeNode> stack = new LinkedList<TreeNode>();
        stack.push(root);
        int inorderIndex = inorder.length - 1;
        for (int i = postorder.length - 2; i >= 0; i--) {
            int postorderVal = postorder[i];
            TreeNode node = stack.peek();
            if (node.val != inorder[inorderIndex]) {
                node.right = new TreeNode(postorderVal);
                stack.push(node.right);
            } else {
                while (!stack.isEmpty() && stack.peek().val == inorder[inorderIndex]) {


                    node = stack.pop();
                    System.out.println(node.val+" "+inorder[inorderIndex]);
                    inorderIndex--;
                }
                node.left = new TreeNode(postorderVal);
                stack.push(node.left);
            }
        }
        return root;
    }

    public int maxDepth(TreeNode root) {
        if(root==null)
            return 0;
        else return 1+Math.max(maxDepth(root.left),maxDepth(root.right));
    }

    boolean ans=true;
    public boolean isBalanced(TreeNode root) {

        getDept(root);
        return ans;

    }

    public int getDept(TreeNode root){
        if(root==null)
            return 0;
        else {
            int n=getDept(root.left);
            int m=getDept(root.right);
            if(Math.abs(n-m)>1)
                ans=false;
            return 1+Math.max(n,m);
        }


    }
    TreeNode treeNode;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

       dfs(root, p, q);

       return treeNode;

    }

    public int[] dfs(TreeNode root,TreeNode p,TreeNode q){
        int a[]=new int[2];
        int b[]=new int[2];
        if(root.left!=null)
        a=dfs(root.left,p,q);
        if(root.right!=null)
        b=dfs(root.right,p,q);

        int ans[]=new int[2];
        if(root.val==p.val)
            ans[0]=1;
        if(root.val==q.val)
            ans[1]=1;
        if(a[0]==1||b[0]==1)
            ans[0]=1;
        if(b[1]==1||a[1]==1)
            ans[1]=1;
        if(ans[0]==1&&ans[1]==1&&treeNode==null)
        {
            treeNode=root;
        }
        return ans;
    }


}
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left,TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week2/day1/Acwing1497/main.go

import "fmt"

/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/25 16:14
 */
const N = 40

var (
    n int
    postorder, inorder [N]int
    l, r, pos = make(map[int]int, 0), make(map[int]int, 0), make(map[int]int, N)
    )

func build(il, ir, pl, pr int) int {
    root := postorder[pr]
    k := pos[root]
    if il < k {l[root] = build(il, k - 1, pl, k - 1 - il + pl)}
    if ir > k {r[root] = build(k + 1, ir, k - 1 - il + pl + 1, pr - 1)}
    return root
}

func bfs(root int) {
    q := []int{root}
    for len(q) > 0 {
        top := q[0]; q = q[1:]
        fmt.Printf("%d ", top)
        if _, ok := l[top]; ok {q = append(q, l[top])}
        if _, ok := r[top]; ok {q = append(q, r[top])}
    }
}

func main() {
    fmt.Scan(&n)
    for i := 0; i < n; i++ {fmt.Scan(&postorder[i])}
    for i := 0; i < n; i++ {
        fmt.Scan(&inorder[i])
        pos[inorder[i]] = i
    }
    root := build(0, n - 1, 0, n - 1)
    bfs(root)
}

